base/bundle_test.cpp

```diff
@@ -0,0 +1,93 @@
+#include "base/bundle.hpp"
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+namespace principia {
+
+using ::testing::Eq;
+
+using namespace std::chrono_literals;  // NOLINT(build/namespace)
+
+namespace base {
+
+class BundleTest : public testing::Test {
+ protected:
+  BundleTest() : bundle_(/*threads=*/8) {}
+
+  Bundle bundle_;
+};
+
+TEST_F(BundleTest, MatrixVectorProduct) {
+  constexpr std::int64_t short_dimension = 100;
+  constexpr std::int64_t long_dimension = 100000;
+  std::vector<std::int64_t> matrix(short_dimension * long_dimension, 1);
+  std::vector<std::int64_t> vector(long_dimension);
+  std::vector<std::int64_t> product(short_dimension);
+  for (int i = 0; i < short_dimension; ++i) {
+    matrix[i + short_dimension * i] = 2;
+  }
+  for (int j = 0; j < long_dimension; ++j) {
+    vector[j] = j;
+  }
+  for (int i = 0; i < short_dimension; ++i) {
+    bundle_.Add(
+        [&matrix, &vector, &product, i, short_dimension, long_dimension]() {
+          product[i] = 0;
+          for (int j = 0; j < long_dimension; ++j) {
+            product[i] += matrix[i + short_dimension * j] * vector[j];
+          }
+          return Status::OK;
+        });
+  }
+  EXPECT_OK(bundle_.Join());
+  for (int i = 0; i < short_dimension; ++i) {
+    EXPECT_THAT(product[i],
+                Eq(long_dimension * (long_dimension - 1) / 2 + i));
+  }
+}
+
+TEST_F(BundleTest, Abort) {
+  int i = 0;
+  auto const wait = [this]() {
+    while (!bundle_.abort_requested()) {}
```

Comment:
Alert!  Ugly busy waiting detected!  Put a sleep in the loop otherwise the test might flake.